File: leetcode/496.下一个更大元素-i.ts
/*
 * @lc app=leetcode.cn id=496 lang=typescript
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
function nextGreaterElement(nums1: number[], nums2: number[]): number[] {
  const nextGreater: Map<number, number> = new Map();
  const stack: number[] = [];

  for (let i = 0, len = nums2.length; i < len; i++) {
    const x = nums2[i];
    while (stack.length > 0 && stack[stack.length - 1] < x) {
      nextGreater.set(stack.pop(), x);
    }
    stack.push(x);
  }

  return nums1.map((num) => (nextGreater.has(num) ? nextGreater.get(num) : -1));
}
// @lc code=end

File: leetcode/496.下一个更大元素-i.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=496 lang=java
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] ans = new int[nums1.length];
        Deque<Integer> stack = new ArrayDeque<>();
        Map<Integer, Integer> map = new HashMap<>();
        for (int x : nums2) {
            while (!stack.isEmpty() && stack.peek() < x) {
                map.put(stack.pop(), x);
            }
            stack.push(x);
        }

        for (int i = 0; i < nums1.length; i++) {
            ans[i] = map.getOrDefault(nums1[i], -1);
        }

        return ans;
    }
}
// @lc code=end
